pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1074 @@
+		Scheduled { id: ScheduleId, who: T::AccountId },
+		///The DCA is planned for blocknumber
+		ExecutionPlanned {
+			id: ScheduleId,
+			who: T::AccountId,
+			block: BlockNumberFor<T>,
+		},
+		///The DCA trade has been successfully executed
+		TradeExecuted { id: ScheduleId, who: T::AccountId },
+		///The DCA trade execution has been failed
+		TradeFailed {
+			id: ScheduleId,
+			who: T::AccountId,
+			error: DispatchError,
+		},
+		///The DCA is terminated and completely removed from the chain
+		Terminated {
+			id: ScheduleId,
+			who: T::AccountId,
+			error: DispatchError,
+		},
+		///The DCA is completed and completely removed from the chain
+		Completed { id: ScheduleId, who: T::AccountId },
+		///Randomness generation failed possibly coming from missing data about relay chain
+		RandomnessGenerationFailed { block: BlockNumberFor<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		///Schedule not exist
+		ScheduleNotFound,
+		///Trade amount is less than fee
+		TradeAmountIsLessThanFee,
+		///Forbidden as the user is not the owner of the schedule
+		Forbidden,
+		///The next execution block number is not in the future
+		BlockNumberIsNotInFuture,
+		///Price change from oracle data is bigger than max allowed
+		PriceChangeIsBiggerThanMaxAllowed,
+		///Error occurred when calculating price
+		CalculatingPriceError,
+		///The total amount to be reserved is smaller than min budget
+		TotalAmountIsSmallerThanMinBudget,
+		///The budget is too low for executing one DCA
+		BudgetTooLow,
+		///There is no free block found to plan DCA execution
+		NoFreeBlockFound,
+		///The DCA schedule has been manually terminated
+		ManuallyTerminated,
+		///Max number of retries reached for schedule
+		MaxRetryReached,
+		///The trade limit has been reached, leading to retry
+		TradeLimitReached,
+		///The route to execute the trade on is not specified
+		RouteNotSpecified,
+		///No parent hash has been found from relay chain
+		NoParentHashFound,
+		///Error that should not really happen only in case of invalid state of the schedule storage entries
+		InvalidState,
+	}
+
+	/// Id sequencer for schedules
+	#[pallet::storage]
+	#[pallet::getter(fn next_schedule_id)]
+	pub type ScheduleIdSequencer<T: Config> = StorageValue<_, ScheduleId, ValueQuery>;
+
+	/// Storing schedule details
+	#[pallet::storage]
+	#[pallet::getter(fn schedules)]
+	pub type Schedules<T: Config> =
+		StorageMap<_, Blake2_128Concat, ScheduleId, Schedule<T::AccountId, T::Asset, BlockNumberFor<T>>, OptionQuery>;
+
+	/// Storing schedule ownership
+	#[pallet::storage]
+	#[pallet::getter(fn owner_of)]
+	pub type ScheduleOwnership<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::AccountId, Twox64Concat, ScheduleId, (), OptionQuery>;
+
+	/// Keep tracking the remaining amounts to spend for DCA schedules
+	#[pallet::storage]
+	#[pallet::getter(fn remaining_amounts)]
+	pub type RemainingAmounts<T: Config> = StorageMap<_, Blake2_128Concat, ScheduleId, Balance, OptionQuery>;
```

Comment:
Schedule entry doesn't have to be ever updated, that is why I recommended using separate storage so that less state has to be rewritten on each execution

Reply:
nothing to do then, leaving it here for @enthusiastmartin for info